frontend/src/lib/Buttons.ts

```diff
@@ -18,112 +18,306 @@ import AddIcon from '@material-ui/icons/Add';
+      outlined: true,
+      title: 'Create recurring run',
+      tooltip: 'Create a new recurring run',
+    };
+  }
+
+  public refresh(action: () => void): ToolbarActionConfig {
+    return {
+      action,
+      id: 'refreshBtn',
+      title: 'Refresh',
+      tooltip: 'Refresh the list',
+    };
+  }
+
+  public upload(action: () => void): ToolbarActionConfig {
+    return {
+      action,
+      icon: AddIcon,
+      id: 'uploadBtn',
+      outlined: true,
+      title: 'Upload pipeline',
+      tooltip: 'Upload pipeline',
+    };
+  }
+
+  private _cloneRun(selectedIds: string[]): void {
+    if (selectedIds.length === 1) {
+      const runId = selectedIds[0];
+      const searchString = this._urlParser.build({ [QUERY_PARAMS.cloneFromRun]: runId || '' });
+      this._props.history.push(RoutePage.NEW_RUN + searchString);
+    }
+  }
+
+  private _deletePipeline(selectedIds: string[], callback: (selectedIds: string[], success: boolean) => void,
+    useCurrent: boolean): void {
+    this._dialogActionHandler(selectedIds, useCurrent,
+      id => Apis.pipelineServiceApi.deletePipeline(id), callback, 'Delete', 'pipeline');
+  }
+
+  private _deleteRecurringRun(id: string, useCurrent: boolean,
+    callback: (_: string[], success: boolean) => void): void {
+    this._dialogActionHandler([id], useCurrent, Apis.jobServiceApi.deleteJob, callback, 'Delete',
+      'recurring run config');
+  }
+
+  private _dialogActionHandler(selectedIds: string[], useCurrent: boolean,
+    api: (id: string) => Promise<void>, callback: (selectedIds: string[], success: boolean) => void,
+    actionName: string, resourceName: string): void {
+
+    const dialogClosedHandler = (confirmed: boolean) =>
+      this._dialogClosed(confirmed, selectedIds, actionName, resourceName, useCurrent, api, callback);
+
+    this._props.updateDialog({
+      buttons: [{
+        onClick: async () => await dialogClosedHandler(true),
+        text: actionName,
+      }, {
+        onClick: async () => await dialogClosedHandler(false),
+        text: 'Cancel',
+      }],
+      onClose: async () => await dialogClosedHandler(false),
+      title: `${actionName} ${useCurrent ? 'this' : selectedIds.length} ${resourceName}${useCurrent ? '' : s(selectedIds.length)}?`,
+    });
+  }
+
+  private async _dialogClosed(confirmed: boolean, selectedIds: string[], actionName: string,
+    resourceName: string, useCurrent: boolean, api: (id: string) => Promise<void>,
+    callback: (selectedIds: string[], success: boolean) => void): Promise<void> {
+    if (confirmed) {
+      const unsuccessfulIds: string[] = [];
+      const errorMessages: string[] = [];
+      await Promise.all(selectedIds.map(async (id) => {
+        try {
+          await api(id);
+        } catch (err) {
+          unsuccessfulIds.push(id);
+          const errorMessage = await errorToMessage(err);
+          errorMessages.push(`Failed to ${actionName.toLowerCase()} ${resourceName}: ${id} with error: "${errorMessage}"`);
```

Comment:
thought: making the action name uppercase would better indicate the generality of it, but maybe we don't want that.

Reply:
Right now it reads nicely as "Failed to delete pipeline: id with error..." which I think is nicer than if it starts with a capital letter.